ow_power_system/src/power_system_node.cpp

```diff
@@ -14,91 +14,131 @@
+    
+    // If voltage limits dip below baseline, set to baseline values
+    if (min_voltage < baseline_voltage)
     {
-      ROS_ERROR_NAMED("power_system_node", "power_update_rate param was set to zero! passed value ignored.");
+      min_voltage = baseline_voltage;
     }
-    else
+    if (max_voltage < (baseline_voltage + voltage_range))
     {
-      power_update_rate = power_update_rate_override;
+      max_voltage = baseline_voltage + voltage_range;
     }
-  }
-
-  ROS_INFO_STREAM_NAMED("power_system_node", "power_update_rate is set to: " << power_update_rate << " Hz");
-  
-  // ROS Loop. Note that once this loop starts,
-  // this function (and node) is terminated with an interrupt.
-  
-  ros::Rate rate(power_update_rate);
-  //individual soc_msg to be published by SOC_pub
-  std_msgs::Int16 rul_msg;
-  rul_msg.data = 0;  // immediately set to a good default
-  std_msgs::Float64 soc_msg;
-  soc_msg.data = 0.0;
-  std_msgs::Float64 tempbat_msg;
-  tempbat_msg.data = 0.0;
-  int temp_index = 1; // Set to 1 for now (constant vector). This will change to median SOC or RUL index or fixed percentile
-  ROS_INFO ("Power system node running");
-
-  while (ros::ok()) {
 
+    // Voltage estimate based on pseudorandom noise and moving range
+    double voltage_est = min_voltage + static_cast <double> (rand()) / RAND_MAX * (max_voltage - min_voltage);
+
+    //individual msgs to be published
+    std_msgs::Int16 rul_msg;
+    rul_msg.data = 0;  // immediately set to a good default
+    std_msgs::Float64 soc_msg;
+    soc_msg.data = 0.0;
+    std_msgs::Float64 tempbat_msg;
+    tempbat_msg.data = 0.0;
+
+    int temp_index = 1; // Set to 1 for now (constant vector). This will change to median SOC or RUL index or fixed percentile
+
+    vector<map<MessageId, Datum<double>>> power_data; // GSAP Input Vector Definition
+    map<MessageId, Datum<double>> initialization_data; // GSAP Initialization Data
+    map<MessageId, Datum<double>> current_data; // GSAP Input Data at Current Timestep
+
+    // Set timestamp for initialization data to initialization time
+    init_power.setTime(init_time);
+    init_temp.setTime(init_time);
+    init_voltage.setTime(init_time);
+
+    // Determine timestamp for current callback
+    auto timestamp = system_clock::now();
+    
+    // Assign input parameters and corresponding timestamp
+    Datum<double> power = mech_power;
+    power.setTime(timestamp);
+    Datum<double> temperature = temp_est;
+    temperature.setTime(timestamp);
+    Datum<double> voltage = voltage_est;
+    voltage.setTime(timestamp);
+
+    // Populate initialization data structure
+    initialization_data.insert({MessageId::Watts, init_power});
+    initialization_data.insert({MessageId::Centigrade, init_temp});
+    initialization_data.insert({MessageId::Volts, init_voltage});
+    power_data.push_back(initialization_data);
```

Comment:
Why do you think that it is necessary to constantly submit the initialization data? I don't understand it .. can you clarify?

Reply:
Just to note that this wasn't the case when we were reading from the csv file .. So I am very skeptic about what is happening in this part of the code.

Reply:
Addressed in latest commit.